server/src/main/java/org/opensearch/cluster/service/TaskBatcher.java

```diff
@@ -90,20 +91,33 @@ public void submitTasks(List<? extends BatchedTask> tasks, @Nullable TimeValue t
                     throw new IllegalStateException("cannot add duplicate task: " + a);
                 }, IdentityHashMap::new));
             LinkedHashSet<BatchedTask> newTasks = new LinkedHashSet<>(tasks);
-            tasksPerBatchingKey.merge(firstTask.batchingKey, newTasks, (existingTasks, updatedTasks) -> {
-                for (BatchedTask existing : existingTasks) {
-                    // check that there won't be two tasks with the same identity for the same batching key
-                    BatchedTask duplicateTask = tasksIdentity.get(existing.getTask());
-                    if (duplicateTask != null) {
-                        throw new IllegalStateException(
-                            "task ["
-                                + duplicateTask.describeTasks(Collections.singletonList(existing))
-                                + "] with source ["
-                                + duplicateTask.source
-                                + "] is already queued"
-                        );
+            taskIdentityPerBatchingKey.merge(
+                firstTask.batchingKey,
+                new LinkedHashSet<>(tasksIdentity.keySet()),
+                (existingIdentity, updatedIdentity) -> {
+                    LinkedHashSet<Object> existingIdentities = taskIdentityPerBatchingKey.computeIfAbsent(
+                        firstTask.batchingKey,
+                        k -> new LinkedHashSet<>(tasksIdentity.keySet().size())
+                    );
+                    for (Object newIdentity : tasksIdentity.keySet()) {
+                        // check that there won't be two tasks with the same identity for the same batching key
+                        if (existingIdentities.contains(newIdentity)) {
+                            BatchedTask duplicateTask = tasksIdentity.get(newIdentity);
+                            throw new IllegalStateException(
+                                "task ["
+                                    + duplicateTask.describeTasks(Collections.singletonList(duplicateTask))
+                                    + "] with source ["
+                                    + duplicateTask.source
+                                    + "] is already queued"
+                            );
+                        }
                     }
+                    existingIdentity.addAll(updatedIdentity);
+                    return existingIdentity;
                 }
+            );
+            // since we have checked for dup tasks in above map, we can add all new task in map.
+            tasksPerBatchingKey.merge(firstTask.batchingKey, newTasks, (existingTasks, updatedTasks) -> {
```

Comment:
Yes I tried that, but after dedup logic we are updating `tasksIdentityPerBatchingKey` map as well, which is not getting updated from scope of merge function of `tasksPerBatchingKey`. 

That's why I had to separate out those two merge function. 

IMO there wont be race condition, because we are updating this two maps in specific order only. Please check more details on order in here: https://github.com/opensearch-project/OpenSearch/pull/7305#discussion_r1178938486

Let me know if I am missing something.